balancer/catalyst/geo.go

```diff
@@ -0,0 +1,56 @@
+package catalyst
+
+import (
+	"math"
+	"sort"
+)
+
+// Earth radius in kilometers
+const earthRadius = 6371
+
+// ToRadians converts degrees to radians
+func toRadians(deg float64) float64 {
+	return deg * (math.Pi / 180)
+}
+
+// Rate the nodes as Good / Okay / Bad based on distance from the request
+func geoScores(nodes []ScoredNode, requestLatitude, requestLongitude float64) []ScoredNode {
+	// Calculate distance from request for each node
+	for i := range nodes {
+		// Convert latitude and longitude from degrees to radians
+		lat1 := toRadians(requestLatitude)
+		lon1 := toRadians(requestLongitude)
+		lat2 := toRadians(nodes[i].GeoLatitude)
+		lon2 := toRadians(nodes[i].GeoLongitude)
+
+		// Haversine formula
+		dlat := lat2 - lat1
+		dlon := lon2 - lon1
+		a := math.Sin(dlat/2)*math.Sin(dlat/2) + math.Cos(lat1)*math.Cos(lat2)*math.Sin(dlon/2)*math.Sin(dlon/2)
+		c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
+
+		// Distance in kilometers
+		nodes[i].GeoDistance = earthRadius * c
+	}
+
+	// Order nodes by distance
+	sort.Slice(nodes, func(i, j int) bool {
+		return nodes[i].GeoDistance < nodes[j].GeoDistance
+	})
+
+	// Assign scores
+	var baseDistance = nodes[0].GeoDistance
+	var goodDistance = baseDistance + 1500
+	var okayDistance = baseDistance + 7500
```

Comment:
Mist approach was different (from what [Jaron described](https://discord.com/channels/423160867534929930/1165251383956602950/1166714774650900532)). It doesn't use hard thresholds but sums different parameters and its weights.

If we want hard if-else logic with thresholds it would be nice to still make it configurable like mist weights so we wouldn't need redeployments for tuning this algorithm

Reply:
Agreed, even if we want this:

```
0km = 2
1500km = 1
7500km = 0
```

No reason not to draw those points on a graph and interpolate between them, yeah? `750km = 1.5` kind of deal